changelog/changelog.py

```diff
@@ -0,0 +1,453 @@
+    """Formats PR number(s) into markdown links like '#pr' joined by ', ' (no angle brackets)"""
+    if prs_data is None or base_url is None: return ""
+    prs_list: List[str] = []
+    match prs_data:
+        case int() | str():
+            pr_str = str(prs_data).strip()
+            if pr_str: prs_list = [pr_str]
+        case list():
+            prs_list = [str(pr).strip() for pr in prs_data if isinstance(pr, (int, str))]
+            prs_list = [pr for pr in prs_list if pr]
+        case _:
+            print(f"⚠️ Unexpected type for 'prs' value: {type(prs_data).__name__} in frontmatter; Skipping PR formatting", file=sys.stderr)
+            return ""
+    if not prs_list: return ""
+    return ", ".join(f"[#{pr}]({base_url}{pr})" for pr in prs_list)
+
+def parse_file(path: Path, pr_base_url: Optional[str]) -> Optional[Tuple[str, str]]:
+    """
+    Parses a single markdown file, extracts frontmatter and content,
+    and validates that singular/plural key pairs (author/authors, pr/prs)
+    do not coexist. Returns None if parsing fails or validation fails
+    """
+    try:
+        post = frontmatter.load(path)
+    except Exception as e:
+        print(f"❌ Error parsing frontmatter in {path.name}: {e}", file=sys.stderr)
+        return None
+
+    has_authors_key = "authors" in post.metadata
+    has_author_key = "author" in post.metadata
+    has_prs_key = "prs" in post.metadata
+    has_pr_key = "pr" in post.metadata
+
+    error_messages = []
+    if has_authors_key and has_author_key:
+        error_messages.append("Found both 'authors' and 'author' keys")
+    if has_prs_key and has_pr_key:
+        error_messages.append("Found both 'prs' and 'pr' keys")
+
+    if error_messages:
+        for msg in error_messages:
+            print(f"❌ Error in {path.name}: {msg}; Please use only one", file=sys.stderr)
+        return None
+
+    raw_authors = post.get("authors", post.get("author"))
+    raw_prs = post.get("prs", post.get("pr"))
+
+    title = str(post.get("title", "Untitled")).strip()
+    type_ = post.get("type")
+    body = post.content.strip()
+
+    if not isinstance(type_, str) or type_ not in SECTION_TITLES:
+        print(f"⚠️ Invalid or unknown type '{type_}' in {path.name}; skipping", file=sys.stderr)
+        return None
+
+    authors_formatted = format_authors(raw_authors)
+    prs_formatted = format_prs(raw_prs, pr_base_url)
+
+    description = f"{body}\n\n"
+    byline = f"By {authors_formatted}" if authors_formatted else ""
+    if prs_formatted:
+        byline += f" in {prs_formatted}." # Period added here
+    elif byline:
+        byline += "." # Period added here if only authors
+    description += byline
+
+    entry = f"### {title}\n\n{description.strip()}\n"
+    return type_, entry
+
+def parse_filename_semver_key(filename_stem: str) -> Tuple:
+    """
+    Creates a sort key for semantic versioning from filenames like 'v5-0-0'
```

Comment:
Try whether you can have a dot in the filename

Reply:
Nope, doesn't work.